Add Kafka Metrics

dilini-muthumala opened the pull request:
## Purpose
Add Kafka Metrics.

## Note 
Certain changes that are in Carbon Analytics v3.0.48-SNAPSHOT are needed for Promethues metrics feature.
This has to be updated to the released version once the Carbon Analytics v3.0.48 is released.

dilini-muthumala committed:
Add WIP Kafka Metrics code

dilini-muthumala committed:
Initial commit

dilini-muthumala committed:
Merge branch 'master' of github.com:dilini-muthumala/siddhi-io-kafka

dilini-muthumala committed:
Add Kafka Metrics code

dilini-muthumala committed:
Add Carbon Analytics 3.0.48-SNAPSHOT dependency

Certain changes that are in 3.0.48-SNAPSHOT are needed for Promethues metrics feature.
This has to be updated to the released version once the v3.0.48 is released.

dilini-muthumala commented:
Note: 
Certain changes that are in Carbon Analytics v3.0.48-SNAPSHOT are needed for Promethues metrics feature.
This has to be updated to the released version once the Carbon Analytics v3.0.48 is released.

senthuran16 commented on component/src/main/java/io/siddhi/extension/io/kafka/metrics/SourceMetrics.java:
```diff
@@ -0,0 +1,98 @@
+            MetricsDataHolder.getInstance().getMetricService()
+                    .gauge(String.format("io.siddhi.SiddhiApps.%s.Siddhi.Kafka.Source.Current.Offset.%s.%s.%s.%s",
+                            siddhiAppName, topic, "partition." + partition, "groupId." + groupId,
+                            "stream_id." + streamId), Level.INFO, () -> topicOffsetMap.get(topic).get(partition));
```
Can `topicOffsetMap.get(topic)` be `null` by any chance, and cause a NPE at `topicOffsetMap.get(topic).get(partition)`?

dilini-muthumala replied:
Made sure that this method is called after updating the topicOffsetMap.

senthuran16 commented on component/src/main/java/io/siddhi/extension/io/kafka/metrics/SinkMetrics.java:
```diff
@@ -0,0 +1,145 @@
+
+    private void updateOffsetMap(String topic, int partition, long offset) {
+        Map<Integer, Long> partitionMap;
+        if (offsetMap.get(topic) == null) {
+            partitionMap = new ConcurrentHashMap();
+        } else {
+            partitionMap = offsetMap.get(topic);
+        }
```
Is `if` block executed only once, and `else` block rest of the time?
If so, shall we swap these conditions?

dilini-muthumala replied:
Could you please elaborate more on this? I think I did not get your point.

senthuran16 replied:
Let's say the following `if` block executes only once:
```java
if (offsetMap.get(topic) == null) {
        partitionMap = new ConcurrentHashMap();
}
```

and thereafter, the `else` block is going to be executed every time.

If that's the case, we can swap the conditions like following:
```java
        if (offsetMap.get(topic) != null) {
            partitionMap = offsetMap.get(topic);
        } else {
            partitionMap = new ConcurrentHashMap();
        }
```

dilini-muthumala replied:
From what I understand, whatever the condition that we put on the if condition will be evaluated. So it's either `offsetMap.get(topic) == null` or `offsetMap.get(topic) != null`
and based on the outcome, one of the following would be exceuted:
- `partitionMap = new ConcurrentHashMap();`
- `partitionMap = offsetMap.get(topic);`

So either way, one condition and one statement will be executed. I do not see a value addition of changing the if-else block as being suggested.

senthuran16 commented on component/src/main/java/io/siddhi/extension/io/kafka/sink/KafkaSink.java:
```diff
@@ -224,10 +243,14 @@ public void publish(Object payload, DynamicOptions dynamicOptions, KafkaSinkStat
             if (null == partitionNo) {
                 producer.send(new ProducerRecord<>(topic, null, System.currentTimeMillis(), key,
-                        payloadToSend));
+                        payloadToSend),
+                        new KafkaProducerCallBack(topic, null, System.currentTimeMillis()));
             } else {
-                producer.send(new ProducerRecord<>(topic, Integer.parseInt(partitionNo), System.currentTimeMillis(),
-                        key, payloadToSend));
+                Integer partition = Integer.parseInt(partitionNo);
+                long currentTime = System.currentTimeMillis();
+                producer.send(new ProducerRecord<>(topic, partition, currentTime,
+                        key, payloadToSend),
+                        new KafkaProducerCallBack(topic, partition, currentTime));
             }
```
Shall we simplify this?
Something like:
```java
Integer partition = (partitionNo != null) ? Integer.parseInt(partitionNo) : null;
long currentTime = System.currentTimeMillis();
producer.send(new ProducerRecord<>(topic, partition, currentTime,
            key, payloadToSend),
            new KafkaProducerCallBack(topic, partition, currentTime));
```

dilini-muthumala replied:
This is executing in the `else` block in which partitionNo is not null.
```
if (null == partitionNo) {
               
} else {
    Integer partition = Integer.parseInt(partitionNo);  //partitionNo is not null since this is executed in the else block
}
```

senthuran16 replied:
The following line will do the null check as you have mentioned.
```java
Integer partition = (partitionNo != null) ? Integer.parseInt(partitionNo) : null;
```
If `partitionNo` is not null, we'll parse it and get a non null `partition`. Otherwise, we'll get a null `partition`.

Then the following lines would function as same as what the existing code does:
```java
long currentTime = System.currentTimeMillis();
producer.send(
            new ProducerRecord<>(
                  topic, 
                  partition,  // either 'null', or a valid Integer (decided by the above code block)
                  currentTime,
                  key, 
                  payloadToSend),
            new KafkaProducerCallBack(
                  topic, 
                  partition, // either 'null', or a valid Integer (decided by the above code block)
                  currentTime));
```

dilini-muthumala replied:
I see. You meant to remove the whole if-else block (That if-else block was already there even before the metrics implementation) 
This is another way of writing that same logic :-) I am not sure what value addition this provides.

senthuran16 commented on component/src/main/java/io/siddhi/extension/io/kafka/source/KafkaConsumerThread.java:
```diff
@@ -72,20 +74,27 @@
     private int trpLength;
+    private SourceMetrics metrics;
+    private String groupId; //Needed for metrics
+//    private String streamId;
```
Shall we remove this unused variable

dilini-muthumala replied:
Removed the commented line. Thanks for pointing that out.

dilini-muthumala committed:
Move update metrics method further down
This is to make sure that the metrics are updated after the TopicOffsetMap is updated.

dilini-muthumala committed:
Check if metrics enabled before updating metrics
